### PR TITLE
[Noise] Add 4d noise, looping, multiple generators.

### PR DESCRIPTION
**Pull request by D8H**
## Changes
* 4d noise
* 2d noise looping (it uses 4d internally)
* multiple named generators with their own settings
* No more choice between Perline or Simplex noise as there is no reason to use Perline noise over Simplex noise and it's a technical choice that can confuse users.

Version 2.0.0 compatibility break:
- Seeds doesn't give the same result as with version 1.x.x

## Example
* Project: https://www.dropbox.com/s/eybx7ytzctk8bq1/marching-squares-island-generator.zip?dl=1
The archive contains 2 projects:
* The island generation untouched example to show compatibility.
* An example to test looping.

image

## JavaScript
It uses another lib to generate noise to have 4d:
* https://github.com/jwagner/simplex-noise.js

```JavaScript

gdjs._extensionNoise = gdjs._extensionNoise || {};

/** Noise generator manager. */
gdjs._extensionNoise.NoiseManager = /** @class */ (function () {
    /**
     * Create the manager of noise generators.
     */
    function NoiseManager() {
        this.seed = gdjs.randomInRange(1, Number.MAX_SAFE_INTEGER);
        /** @type {Map<string, gdjs._extensionNoise.NoiseGenerator>} */
        this.generators = new Map();
    }

    /**
     * @param name {string}
     * @return {gdjs._extensionNoise.NoiseGenerator}
     */
    NoiseManager.prototype.getGenerator = function (name) {
        let generator = this.generators.get(name);
        if (!generator) {
            generator = new gdjs._extensionNoise.NoiseGenerator(name + this.seed);
            this.generators.set(name, generator);
        }
        return generator;
    }

    /**
     * @param seed {number}
     */
    NoiseManager.prototype.setSeed = function (seed) {
        this.seed = seed;
        this.generators.forEach(generator => generator.setSeed(seed));
    }

    /**
     * @param name {string}
     */
    NoiseManager.prototype.deleteGenerator = function (name) {
        this.generators.delete(name);
    }

    /**
     */
    NoiseManager.prototype.deleteAllGenerators = function () {
        this.generators.clear();
    }

    return NoiseManager;
}());

/** Noise generator with octaves. */
gdjs._extensionNoise.NoiseGenerator = /** @class */ (function () {
    /**
     * Create a noise generator with a seed.
     * @param seed {string}
     */
    function NoiseGenerator(seed) {
        this.simplexNoise = new gdjs._extensionNoise.SimplexNoise(seed);
        this.frequency = 1;
        this.octaves = 1;
        this.persistence = 0.5;
        this.lacunarity = 2;
        this.xLoopPeriod = 0;
        this.yLoopPeriod = 0;
    }

    /**
     * @param seed {string}
     */
    NoiseGenerator.prototype.setSeed = function(seed) {
        this.simplexNoise = new gdjs._extensionNoise.SimplexNoise(seed);
    }

    /**
     * @param x {float}
     * @param y {float}
     * @param z {float} optionnal
     * @param w {float} optionnal
     * @return {float}
     */
    NoiseGenerator.prototype.noise = function (x, y, z, w) {
        if (this.xLoopPeriod && this.yLoopPeriod) {
            const circleRatioX = 2 * Math.PI / this.xLoopPeriod;
            const circleRatioY = 2 * Math.PI / this.yLoopPeriod;
            const angleX = circleRatioX * x;
            const angleY = circleRatioY * y;
            x = Math.cos(angleX) / circleRatioX;
            y = Math.sin(angleX) / circleRatioX;
            z = Math.cos(angleY) / circleRatioY;
            w = Math.sin(angleY) / circleRatioY;
        }
        else if (this.xLoopPeriod) {
            const circleRatio = 2 * Math.PI / this.xLoopPeriod;
            const angleX = circleRatio * x;
            w = z;
            z = y; 
            x = Math.cos(angleX) / circleRatio;
            y = Math.sin(angleX) / circleRatio;
        }
        else if (this.yLoopPeriod) {
            const circleRatio = 2 * Math.PI / this.xLoopPeriod;
            const angleX = circleRatio * x;
            w = z;
            // Make the circle perimeter equals to the looping period
            // to keep the same perceived frequency with or without looping.
            y = Math.cos(angleX) / circleRatio;
            z = Math.sin(angleX) / circleRatio;
        }
        let noiseFunction = this.simplexNoise.noise4D.bind(this.simplexNoise);
        if (z === undefined) {
            noiseFunction = this.simplexNoise.noise2D.bind(this.simplexNoise);
        }
        else if (w === undefined) {
            noiseFunction = this.simplexNoise.noise3D.bind(this.simplexNoise);
        }
        let frequency = this.frequency;
        let noiseSum = 0;
        let amplitudeSum = 0;
        let amplitude = 1;
        for (let i = 0; i < this.octaves; i++) {
            noiseSum += noiseFunction(x * frequency, y * frequency, z * frequency, w * frequency) * amplitude;
            amplitudeSum += Math.abs(amplitude);
            amplitude *= this.persistence;
            frequency *= this.lacunarity;
        }
        return noiseSum / amplitudeSum;
    }

    return NoiseGenerator;
}());
```

### CODE REVIEW

**Comment by tristanbob** [comment]
I can't comment on the javascript due to my limited skill set, but I did review and test this update.

- Using deprecated version of "Simplex2" works
- Using new version of "Noise2D" works
- Text and descriptions look good

Questions:

1) I get different results by changing the "Name" of the generator.  Is this expected?  

2) What is the purpose of requiring a name for the generator?

Video:

https://user-images.githubusercontent.com/8879811/159177159-7cad1421-306c-4836-b835-bbd5676064f2.mp4

**Comment by D8H** [comment]
>     1. I get different results by changing the "Name" of the noise generator.  Is this expected?

Yes, internally the actual seed of a generator is the generator name suffixed by the seed number. 

>     2. What is the purpose of requiring a name for the noise generator?

Let's say you want to generate position of flowers and trees. 2 generators can be used to have different positions without relying on big offset (big offsets can still be used if memory is an issue).

>     3. Should the seed be set per noise generator?

Having a global seed alows to generate the exact same world with this seed (if generator names stays the same)

File: Extensions/Noise.json [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
   "extensionNamespace": "",
   "fullName": "Noise generator",
   "helpPath": "/tutorials/procedural-generation",
   "iconUrl": "data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz48IURPQ1RZUEUgc3ZnIFBVQkxJQyAiLS8vVzNDLy9EVEQgU1ZHIDEuMS8vRU4iICJodHRwOi8vd3d3LnczLm9yZy9HcmFwaGljcy9TVkcvMS4xL0RURC9zdmcxMS5kdGQiPjxzdmcgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayIgdmVyc2lvbj0iMS4xIiBpZD0ibWRpLXNwcmlua2xlciIgd2lkdGg9IjI0IiBoZWlnaHQ9IjI0IiB2aWV3Qm94PSIwIDAgMjQgMjQiPjxwYXRoIGQ9Ik0xMSA3SDEzVjlIMTFWN001IDIySDlWMTBINVYyMk0xNCAxMUgxNlY5SDE0VjExTTE3IDEwSDE5VjhIMTdWMTBNMTcgNVY3SDE5VjVIMTdNMTQgOEgxNlY2SDE0VjhNMTcgMTNIMTlWMTFIMTdWMTNNNSA3SDUuMzNMNiA5SDhMOC42NyA3SDlWNkg1VjdaIiAvPjwvc3ZnPg==",
   "name": "Noise",
   "previewIconUrl": "https://resources.gdevelop-app.com/assets/Icons/sprinkler.svg",
-  "shortDescription": "Expressions for generating Perlin and Simplex noise values (2D or 3D). Useful for procedural generation.",
-  "version": "1.2.0",
+  "shortDescription": "Expressions for generating Perlin and Simplex noise values. Useful for procedural generation.",
```
Comment (tristanbob) [inline Extensions/Noise.json:11]:
```suggestion
  "shortDescription": "Expressions for generating Simplex noise values. Useful for procedural generation.",
```

**Review by tristanbob — APPROVED** [review]
I tested 2D noise on Colony Crawler and it works well (both deprecated and new functions).

I removed "Perlin" from the description since that is deprecated.  However, I left it as a tag since people might be searching for that term.

**Comment by Add00** [comment]
I'm not the best programmer, but the code looks good to me. I think depreciating the perlin functions was the right call since simplex is better in all ways. Only thing that I say is maybe the map function should be included inside of the noise extension since in almost all cases noise values have to be mapped to be useful. Thanks so much for taking the time to improve the noise extension and making it better for all of us.

**Comment by tristanbob** [comment]
> Only thing that I say is maybe the map function should be included inside of the noise extension since in almost all cases noise values have to be mapped to be useful. 

Have you been looking at my code?  ;)  I agree that the Map function is VERY valuable with noise, but I don't think we should include it.  However, we could add a tip in the description so people know how to access it (ExtendedMath extension).

image

> Thanks so much for taking the time to improve the noise extension and making it better for all of us.

Yes, I agree!  Good work.

**Comment by tristanbob** [comment]
Since this has passed two reviews, I think it can be merged.  

However, I have a Github process question:   Should I accept the default "Squash and merge" or one of the other options?

**Comment by 4ian** [comment]
You can squash and merge, editing the title and description so that it will be readable in the changelog. 

You can either remove the description or write bullet points. Example:

Title: `Add extension Blabla to do something`
Description: 
```
* This is useful for balboa
* And this too
```

**Comment by D8H** [comment]
I added this sentence in noise expressions: The "Map" expression from Extended Math extension can be used to map values to any chosen bounds.

**Pull request merged**